src/verify.c

```diff
@@ -348,125 +348,128 @@ static int get_all_files(struct manifest *official_manifest, struct list *subs)
 	return 0;
 }
 
-static struct list *download_loop(struct list *files, int isfailed)
+static struct list *download_loop(struct list *files, bool free_list)
 {
 	int ret;
-	struct file local;
 	struct list *iter;
 	unsigned int complete = 0;
 	unsigned int list_length = list_len(files);
 
 	iter = list_head(files);
 	while (iter) {
 		struct file *file;
-		char *fullname;
 
 		file = iter->data;
 		iter = iter->next;
 		complete++;
 
-		if (file->is_deleted) {
+		if (file->is_deleted || file->do_not_update) {
 			continue;
 		}
 
-		fullname = mk_full_filename(path_prefix, file->filename);
-		if (fullname == NULL) {
-			abort();
-		}
-
-		memset(&local, 0, sizeof(struct file));
-		local.filename = file->filename;
-		populate_file_struct(&local, fullname);
-
-		if (cmdline_option_quick) {
-			ret = compute_hash_lazy(&local, fullname);
-		} else {
-			ret = compute_hash(&local, fullname);
-		}
-		if (ret != 0) {
-			free_string(&fullname);
-			continue;
-		}
-
-		if (hash_needs_work(file, local.hash)) {
-			/* Mix content is local, so don't queue files up for curl downloads */
-			if (file->is_mix) {
-				char *filename;
-				char *url;
-				string_or_die(&url, "%s/%i/files/%s.tar", MIX_STATE_DIR, file->last_change, file->hash);
-				string_or_die(&filename, "%s/download/.%s.tar", state_dir, file->hash);
-				file->staging = filename;
-				ret = link(url, filename);
-				/* Try doing a regular rename if hardlink fails */
-				if (ret) {
-					if (rename(url, filename) != 0) {
-						fprintf(stderr, "Failed to copy local mix file: %s\n", filename);
-						continue;
-					}
+		/* Mix content is local, so don't queue files up for curl downloads */
+		if (file->is_mix) {
+			char *filename;
+			char *url;
+			string_or_die(&url, "%s/%i/files/%s.tar", MIX_STATE_DIR, file->last_change, file->hash);
+			string_or_die(&filename, "%s/download/.%s.tar", state_dir, file->hash);
+			file->staging = filename;
+			ret = link(url, filename);
+			/* Try doing a regular rename if hardlink fails */
+			if (ret) {
+				if (rename(url, filename) != 0) {
+					fprintf(stderr, "Failed to copy local mix file: %s\n", filename);
+					continue;
 				}
-				untar_full_download(file);
-				free_string(&filename);
-				free_string(&url);
-				continue;
 			}
-			full_download(file);
-		} else {
-			/* mark the file as good to save time later */
-			file->do_not_update = 1;
+			untar_full_download(file);
+			free_string(&filename);
+			free_string(&url);
+			continue;
```

Comment:
Hmm, not your fault, but I notice we aren't properly printing progress for our mixed files.

Reply:
Good point. I'll add that in a future commit.